File: auto-linkedin.js
const puppeteer = require("puppeteer");
const fs = require("fs");
const XLSX = require("xlsx");

const chromedriver_path = "path_to_chromedriver";

async function loginLinkedIn(page, email, password) {
  await page.goto("https://www.linkedin.com/login");
  await page.type('input[name="session_key"]', email);
  await page.type('input[name="session_password"]', password);
  await page.click('button[type="submit"]');
  await page.waitForNavigation();
}

async function searchLinkedInProfile(page, name) {
  const searchQuery = `${name} organization_name`;

  await page.waitForSelector(".search-global-typeahead__input");
  const searchBox = await page.$(".search-global-typeahead__input");
  await searchBox.click({ clickCount: 3 });
  await searchBox.type(searchQuery);
  await searchBox.press("Enter");
  await page.waitForTimeout(2000);

  const noResults = await page.$x("//h2[text()='No results found']");
  if (noResults.length) {
    console.log("No results found for:", name);
    return null;
  }

  const noAccess = await page.$x(
    "//h2[@id='out-of-network-modal-header' and text()='You don’t have access to this profile']"
  );
  if (noAccess.length) {
    console.log("No access to profile for:", name);
    const gotItButton = await page.$x("//button[contains(., 'Got it')]");
    await page.waitForTimeout(2000);
    await gotItButton[0].click();
    await page.waitForTimeout(4000);
    return null;
  }

  const firstResult = await page.$x(
    '//span[contains(@class, "entity-result__title-text")]//a[contains(@class, "app-aware-link")]'
  );
  await firstResult[0].click();
  await page.waitForTimeout(2000);

  const currentUrl = page.url();
  if (currentUrl.includes("linkedin.com/jobs")) {
    await page.goto("https://www.linkedin.com/feed/");
    console.log("No results found for:", name);
    console.log("Current page is LinkedIn Jobs. Redirecting to LinkedIn feed.");
    return null;
  } else if (currentUrl.includes("linkedin.com/in")) {
    console.log("Profile URL:", currentUrl);
    return currentUrl;
  }

  return null;
}

(async () => {
  const workbook = XLSX.readFile("load_file.xlsx");
  const sheetName = workbook.SheetNames[0];
  const sheet = workbook.Sheets[sheetName];
  const data = XLSX.utils.sheet_to_json(sheet, { header: 1 });
  const headers = data[0];
  const rows = data.slice(1);

  const browser = await puppeteer.launch({
    headless: false,
    args: ["--ignore-certificate-errors", "--ignore-ssl-errors"],
  });
  const page = await browser.newPage();
  page.setDefaultNavigationTimeout(60000);

  await loginLinkedIn(page, "email", "password");

  for (const [index, row] of rows.entries()) {
    const name = row[headers.indexOf("NAME")];

    if (!name) {
      console.log(`Name at index ${index} is empty or not valid.`);
      continue;
    }

    const profileUrl = await searchLinkedInProfile(page, name);
    if (profileUrl) {
      row[headers.indexOf("LINK")] = profileUrl;
    }
  }

  const newSheet = XLSX.utils.aoa_to_sheet([headers, ...rows]);
  const newWorkbook = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(newWorkbook, newSheet, sheetName);
  XLSX.writeFile(newWorkbook, "save_to_file.xlsx");

  await browser.close();
})();

File: auto-linkedin.py
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import NoSuchElementException
import pandas as pd
import time

chromedriver_path = 'path_to_chromedriver'

def login_linkedin(email, password):
    driver.get('https://www.linkedin.com/login')
    driver.find_element(By.NAME, 'session_key').send_keys(email)
    driver.find_element(By.NAME, 'session_password').send_keys(password)
    driver.find_element(By.XPATH, "//button[contains(text(), 'Sign in')]").click() 

def search_linkedin_profile(name):
    search_query = f'{name} organization_name' 
    search_box = WebDriverWait(driver, 100).until(EC.element_to_be_clickable((By.CLASS_NAME, "search-global-typeahead__input")))
    search_box.clear()
    search_box.send_keys(search_query)
    search_box.send_keys(Keys.RETURN)

    time.sleep(2)

    try:
        no_results_element = driver.find_element(By.XPATH, "//h2[text()='No results found']")
        print("No results found for:", name)
        return
    except NoSuchElementException:
        pass

    try:
        no_access_element = driver.find_element(By.XPATH, "//h2[@id='out-of-network-modal-header' and text()='You don’t have access to this profile']")
        print("No access to profile for:", name)
        got_it_button = driver.find_element(By.XPATH, "//button[contains(., 'Got it')]")
        time.sleep(2)
        got_it_button.click()
        time.sleep(4)
        return
    except NoSuchElementException:
        pass
    
    first_search_result = WebDriverWait(driver, 100).until(EC.element_to_be_clickable((By.XPATH, '//span[contains(@class, "entity-result__title-text")]//a[contains(@class, "app-aware-link")]')))
    first_search_result.click()
    
    time.sleep(2)

    current_url = driver.current_url

    if "linkedin.com/jobs" in current_url:
        driver.get("https://www.linkedin.com/feed/")
        print("No results found for:", name)
        print("Current page is LinkedIn Jobs. Redirecting to LinkedIn feed.")
    elif "linkedin.com/in" in current_url:
        profile_url = driver.current_url
        print("Profile URL:", profile_url)

excel_file = 'load_file.xlsx'
df = pd.read_excel(excel_file)

service = Service(executable_path=chromedriver_path)
options = Options()
options.add_argument('--ignore-certificate-errors')
options.add_argument('--ignore-ssl-errors')
driver = webdriver.Chrome(service=service, options=options)
driver.maximize_window()

login_linkedin('email', 'password')

for index, row in df.iterrows():
    name = row['NAME']

    if pd.isna(name) or name == "":
        print(f"Name at index {index} are empty or not valid.")
        continue
    
    search_linkedin_profile(name)
    profile_url = driver.current_url
    df.at[index, 'LINK'] = profile_url

df.to_excel('save_to_file.xlsx', index=False)

driver.quit()